Neos.Neos/Classes/FrontendRouting/Projection/DocumentUriPathFinder.php

```diff
@@ -257,9 +257,63 @@ private function fetchSingle(string $where, array $parameters): DocumentNodeInfo
         return DocumentNodeInfo::fromDatabaseRow($row);
     }
 
+    /**
+     * @param string $where
+     * @param array<string> $parameters
+     * @return DocumentNodeInfo[]
+     */
+    private function fetchMultiple(string $where, array $parameters): array
+    {
+        try {
+            $rows = $this->dbal->fetchAllAssociative(
+                'SELECT * FROM ' . $this->tableNamePrefix . '_uri
+                     WHERE ' . $where,
+                $parameters,
+                DocumentUriPathProjection::COLUMN_TYPES_DOCUMENT_URIS
+            );
+        } catch (DBALException $e) {
+            throw new \RuntimeException(sprintf(
+                'Failed to load node for query "%s": %s',
+                $where,
+                $e->getMessage()
+            ), 1683808640, $e);
+        }
+
+        return array_map(fn($row) => DocumentNodeInfo::fromDatabaseRow($row), $rows);
+    }
+
+    private function getCacheKey(NodeAggregateId $nodeAggregateId, string $dimensionSpacePointHash): string
+    {
+        return $nodeAggregateId->value . '#' . $dimensionSpacePointHash;
+    }
+
+    public function purgeCacheFor(DocumentNodeInfo $nodeInfo): void
+    {
+        if ($this->cacheEnabled) {
+            $cacheKey = $this->getCacheKey($nodeInfo->getNodeAggregateId(), $nodeInfo->getDimensionSpacePointHash());
+            unset($this->getByIdAndDimensionSpacePointHashCache[$cacheKey]);
+        }
+    }
+
     public function disableCache(): void
     {
         $this->cacheEnabled = false;
         $this->getByIdAndDimensionSpacePointHashCache = [];
     }
+
+    /**
+     * @param DocumentNodeInfo $node
+     * @return DocumentNodeInfo[]
+     */
+    public function getAllChildrenOfNode(DocumentNodeInfo $node): array
```

Comment:
"Children" is not the correct terminology since this works recursively.
Also I would suggest to avoid `array` types in APIs.
If it does not exist yet, we could introduce a generic `DocumentNodeInfos` iterator
```suggestion
    public function getDescendantsOfNode(DocumentNodeInfo $node): DocumentNodeInfos
```